moped-editor/src/views/projects/projectView/ProjectComponentEdit.js

```diff
@@ -0,0 +1,777 @@
+    features: [],
+  };
+
+  /**
+   * The State
+   * @type {String} selectedComponentType - A string containing the name of the selected type in lowercase
+   * @type {String} selectedComponentSubtype - A string containing the name of the selected subtype in lowercase
+   * @type {String[]} selectedComponentSubtype - A string list containing all available subtypes for type
+   * @type {Number[]} selectedSubcomponents - A number array containing the id of the selected subcomponents
+   * @type {String[]} availableSubtypes - An string list containing all available subtypes for the selected component
+   * @type {Object[]} editFeatureComponents - An object list containing the features for this component
+   * @type {Object} editFeatureCollection - The final GeoJson generated for all the the features in this component
+   * @type {String} componentDescription - The description of this component
+   * @type {boolean} deleteDialogOpen - If true, it displays the delete dialog, or hides it if false.
+   * @constant
+   */
+  const [selectedComponentId, setSelectedComponentId] = useState(null);
+  const [selectedComponentType, setSelectedComponentType] = useState(null);
+  const [selectedComponentSubtype, setSelectedComponentSubtype] = useState(
+    null
+  );
+  const [selectedSubcomponents, setSelectedSubcomponents] = useState([]);
+  const [availableSubtypes, setAvailableSubtypes] = useState([]);
+  const [editFeatureComponents, setEditFeatureComponents] = useState([]);
+  const [editFeatureCollection, setEditFeatureCollection] = useState(
+    emptyFeatureCollection
+  );
+
+  const [componentDescription, setComponentDescription] = useState(null);
+
+  const [deleteDialogOpen, setDeleteDialogOpen] = useState(false);
+
+  /**
+   * Apollo hook functions
+   */
+  const { loading, data, error } = useQuery(COMPONENT_DETAILS_QUERY, {
+    variables: {
+      componentId: componentId,
+    },
+    fetchPolicy: "no-cache",
+  });
+
+  const [updateProjectComponents] = useMutation(UPDATE_MOPED_COMPONENT);
+
+  const [deleteProjectComponent] = useMutation(DELETE_MOPED_COMPONENT);
+
+  /**
+   * Generates an initial list of component types, subtypes and counts
+   */
+  const initialTypeCounts = data
+    ? data.moped_components.reduce((accumulator, component, index) => {
+        const componentId = component?.component_id ?? null;
+        const componentName = (component?.component_name ?? "").toLowerCase();
+        const componentSubtype = (
+          component?.component_subtype ?? ""
+        ).toLowerCase();
+
+        const componentSubcomponents = data.moped_subcomponents.filter(
+          subcomponent => subcomponent.component_id === componentId
+        );
+
+        // Get the total count
+        const currentCount = accumulator[componentName]?.count ?? 0;
+
+        return {
+          ...accumulator,
+          [componentName]: {
+            count: currentCount + 1,
+            component_id: currentCount > 1 ? 0 : componentId,
```

Comment:
I am very much still trying to wrap my head around components, sub components etc -- why is the component_id `0` if there is at least one subtype?

Reply:
@chiaberry sorry for the late reply I had to check. I remember now, `currentCount > 1 ` forces the id to be zero whenever we have a group of type-subtype pairs. Sometimes, we only have a type without a subtype (see "bike box" in the data below), and the one does have a component_id because it is unique (not a group).

This is a copy of the output object, in other words this is what `initialTypeCounts` stores after this reduce function executes:

```
const initialTypeCounts = {
    "access control": {
        "count": 4,
        "component_id": 0,
        "index": 3,
        "subtypes": {
            "driveway closure": {
                "component_id": 1,
                "component_name": "Access Control",
                "component_subtype": "Driveway Closure"
            },
            "driveway modification": {
                "component_id": 2,
                "component_name": "Access Control",
                "component_subtype": "Driveway Modification"
            },
            "hardened centerline": {
                "component_id": 3,
                "component_name": "Access Control",
                "component_subtype": "Hardened Centerline"
            },
            "median": {
                "component_id": 4,
                "component_name": "Access Control",
                "component_subtype": "Median"
            }
        },
        "subcomponents": [
            {
                "subcomponent_name": "Leading pedestrian interval",
                "subcomponent_id": 4,
                "component_id": 4,
                "__typename": "moped_subcomponents"
            }
        ]
    },
    "bike lane": {
        "count": 7,
        "component_id": 0,
        "index": 17,
        "subtypes": {
            "buffered": {
                "component_id": 7,
                "component_name": "Bike Lane",
                "component_subtype": "Buffered"
            },
            "climbing": {
                "component_id": 8,
                "component_name": "Bike Lane",
                "component_subtype": "Climbing"
            },
            "colored": {
                "component_id": 9,
                "component_name": "Bike Lane",
                "component_subtype": "Colored"
            },
            "contraflow": {
                "component_id": 10,
                "component_name": "Bike Lane",
                "component_subtype": "Contraflow"
            },
            "protected one-way": {
                "component_id": 11,
                "component_name": "Bike Lane",
                "component_subtype": "Protected One-Way"
            },
            "protected two-way": {
                "component_id": 12,
                "component_name": "Bike Lane",
                "component_subtype": "Protected Two-Way"
            },
            "": {
                "component_id": 6,
                "component_name": "Bike Lane",
                "component_subtype": null
            }
        },
        "subcomponents": []
    },
    "crossswalk": {
        "count": 3,
        "component_id": 0,
        "index": 12,
        "subtypes": {
            "continental": {
                "component_id": 13,
                "component_name": "Crossswalk",
                "component_subtype": "Continental"
            },
            "creative": {
                "component_id": 14,
                "component_name": "Crossswalk",
                "component_subtype": "Creative"
            },
            "raised": {
                "component_id": 15,
                "component_name": "Crossswalk",
                "component_subtype": "Raised"
            }
        },
        "subcomponents": []
    },
    "signal": {
        "count": 3,
        "component_id": 0,
        "index": 15,
        "subtypes": {
            "phb": {
                "component_id": 16,
                "component_name": "Signal",
                "component_subtype": "PHB"
            },
            "rrfb": {
                "component_id": 17,
                "component_name": "Signal",
                "component_subtype": "RRFB"
            },
            "traffic": {
                "component_id": 18,
                "component_name": "Signal",
                "component_subtype": "Traffic"
            }
        },
        "subcomponents": []
    },
    "bike box": {
        "count": 1,
        "component_id": 5,
        "index": 16,
        "subtypes": {
            "": {
                "component_id": 5,
                "component_name": "Bike Box",
                "component_subtype": null
            }
        },
        "subcomponents": []
    },
    "project extent - generic": {
        "count": 1,
        "component_id": 0,
        "index": 18,
        "subtypes": {
            "": {
                "component_id": 0,
                "component_name": "Project Extent - Generic",
                "component_subtype": null
            }
        },
        "subcomponents": []
    }
}
```